File: src/client/java/fr/hardel/asset_editor/client/javafx/components/ui/ItemSprite.java
package fr.hardel.asset_editor.client.javafx.components.ui;

import fr.hardel.asset_editor.client.javafx.lib.ItemAtlasGenerator;
import fr.hardel.asset_editor.client.rendering.ItemAtlasRenderer.AtlasEntry;
import javafx.geometry.Rectangle2D;
import javafx.scene.image.ImageView;
import javafx.scene.image.WritableImage;
import net.minecraft.resources.Identifier;

public final class ItemSprite extends ImageView {
    private final Identifier itemId;
    private Runnable subscription;

    public ItemSprite(Identifier itemId, double displaySize) {
        this.itemId = itemId;
        setFitWidth(displaySize);
        setFitHeight(displaySize);
        setPreserveRatio(true);
        setSmooth(false);

        sceneProperty().addListener((obs, oldScene, newScene) -> {
            if (newScene == null) {
                if (subscription != null) {
                    subscription.run();
                    subscription = null;
                }
                return;
            }

            if (subscription == null)
                subscription = ItemAtlasGenerator.subscribe(this::refresh);
            refresh();
        });

        refresh();
    }

    private void refresh() {
        WritableImage atlas = ItemAtlasGenerator.getAtlasImage();
        AtlasEntry entry = ItemAtlasGenerator.getEntry(itemId);
        if (atlas == null || entry == null) {
            setVisible(false);
            setManaged(false);
            setImage(null);
            return;
        }

        setVisible(true);
        setManaged(true);
        setImage(atlas);
        setViewport(new Rectangle2D(entry.x(), entry.y(), entry.size(), entry.size()));
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/components/ui/ItemSprite.kt
package fr.hardel.asset_editor.client.compose.components.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.FilterQuality
import androidx.compose.ui.unit.Dp
import fr.hardel.asset_editor.client.compose.lib.ItemAtlasGenerator
import kotlin.math.roundToInt
import net.minecraft.resources.Identifier

@Composable
fun ItemSprite(
    itemId: Identifier,
    displaySize: Dp,
    modifier: Modifier = Modifier
) {
    val atlas by ItemAtlasGenerator.atlasImageState
    val currentGeneration by ItemAtlasGenerator.generation

    LaunchedEffect(atlas == null) {
        if (atlas == null) ItemAtlasGenerator.getAtlasImage()
    }

    val image = atlas ?: return
    val entry = remember(itemId, currentGeneration) { ItemAtlasGenerator.getEntry(itemId) } ?: return

    Canvas(
        modifier = modifier.size(displaySize)
    ) {
        drawImage(
            image = image,
            srcOffset = androidx.compose.ui.unit.IntOffset(entry.x(), entry.y()),
            srcSize = androidx.compose.ui.unit.IntSize(entry.size(), entry.size()),
            dstSize = androidx.compose.ui.unit.IntSize(size.width.roundToInt(), size.height.roundToInt()),
            filterQuality = FilterQuality.None
        )
    }
}
